### DISCUSSION TITLE
Inline snapshots?

### DISCUSSION THREAD

**Discussion by lydell** [avajs/ava#2816] [Ideas]
Jest has [inline snapshots](https://jestjs.io/docs/snapshot-testing#inline-snapshots). I quite like them for testing strings, such as error messages and CLI output.

- I can just write `expect(errorMessage).toMatchInlineSnapshot()`. No fiddly copy-pasting and manual typing of the expected value.
- Jest automatically fills in the actual value inside the `()`.
- I look at the filled in value if it looks ok.
- If I ever tweak that error message, I can just tell Jest to update the snapshot.
- I get to see both the test code and the expected output in one place (don’t need to jump to a separate snapshot file).
- Inline snapshots feels a lot like testing your code by running stuff in the REPL and looking at the results (easy and fast), with the big bonus that you can re-run all tests later!

Here’s an example from a project I’m working on:

```js
    test("empty outputs", async () => {
      expect(await run("elm-tooling-json-decode-error/empty-outputs", ["make"]))
        .toMatchInlineSnapshot(`
        ⧙-- INVALID elm-tooling.json FORMAT ---------------------------------------------⧘
        /Users/you/project/tests/fixtures/errors/elm-tooling-json-decode-error/empty-outputs/elm-tooling.json

        I read inputs, outputs and options from ⧙elm-tooling.json⧘.

        ⧙I had trouble with the JSON inside:⧘

        At root["x-elm-watch"]["outputs"]:
        Expected a non-empty object
        Got: {}
      `);
    });
```

Has anything like this ever been discussed in AVA? I tried to search but couldn’t find anything. Is this something there would be interest in?

I like this feature of Jest very much, but I also like the leanness of AVA!

Two random references (not sure if they help at all):

- Jest’s code (`jest-snapshot` package): https://github.com/facebook/jest/tree/master/packages/jest-snapshot
- An assertion library implementing inline snapshots (haven’t tried it): https://github.com/unexpectedjs/unexpected-snapshot

**Comment by novemberborn** [comment]
AVA's snapshots are not text based, so I think this would become quite awkward. We'd have to put an additional base64-encoded segment in there.
